### PR TITLE
Add Kernel.tap/2

### PR DESCRIPTION
**Pull request by v0idpwn**
Missing tests and examples, will be adding soon. Does half of #10602

### CODE REVIEW

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -1126,6 +1126,17 @@ defmodule Kernel do
     :erlang.spawn_monitor(module, fun, args)
   end
 
+  @doc """
+  Applies a function over a term and returns term
+
+  Useful for running side effects
+  """
+  @spec tap(x :: any, (x -> any)) :: x
```
Comment (Eiji7) [inline lib/elixir/lib/kernel.ex:1134]:
```suggestion
  @spec tap(x, (x -> any)) :: x when x: any
```

**Review by Eiji7 — CHANGES_REQUESTED** [review]

**Comment by Eiji7** [comment]
@v0idpwn You should add also:
```elixir
@doc since: "1.12.0"
```
and some examples in documentation and test (check my PR for a reference).

File: lib/elixir/lib/kernel.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/kernel.ex]: Use “var” instead of any. Also please add examples, @doc since, and tests.

File: lib/elixir/lib/kernel.ex [hunk 2]
```diff
@@ -1126,6 +1126,17 @@ defmodule Kernel do
     :erlang.spawn_monitor(module, fun, args)
   end
 
+  @doc """
+  Applies a function over a term and returns term
+
+  Useful for running side effects
+  """
+  @spec tap(x, (x -> any)) :: x when x: term
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:1134]: Use “var” instead of any. Also please add examples, @doc since, and tests.

Comment (josevalim) [reply lib/elixir/lib/kernel.ex:1134]: Also please use “value” instead of “term” and add periods at the end of sentences.

Comment (v0idpwn) [reply lib/elixir/lib/kernel.ex:1134]: Sorry, @josevalim, I opened the PR before finishing just to signalize work was being done

**Comment by v0idpwn** [comment]
Just pushed a first version. 

I need to use some kind of side effect on the doctests, but using IO.inspect() will print during tests. Do you have any other idea of illustrative tests which would serve for this purpose?

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -1126,6 +1126,27 @@ defmodule Kernel do
...
+  Useful for running synchronous side effects when you don't want their
+  result to affect a pipe chain.
+
+  ## Examples
+
+    iex> tap(1, fn x -> x + 1 end)
+    1
+
+    iex> tap(%{a: 1}, &IO.inspect(Map.keys(&1)))
+    %{a: 1}
```
Comment (v0idpwn) [inline lib/elixir/lib/kernel.ex:1141]: This doctest prints during tests. I want a side-effect that doesn't print but could be useful for the reader. Nothing comes in mind.

Comment (maennchen) [reply lib/elixir/lib/kernel.ex:1141]: How about a `send/2` to somewhere?

File: lib/elixir/lib/kernel.ex [hunk 4]
```diff
@@ -1126,6 +1126,27 @@ defmodule Kernel do
     :erlang.spawn_monitor(module, fun, args)
   end
 
+  @doc """
+  Applies a function over a value returns the initial value.
+
+  Useful for running synchronous side effects when you don't want their
+  result to affect a pipe chain.
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:1133]:
```suggestion
  Pipes `value` to the given `fun` and returns the `value` itself.

  Useful for running synchronous side effects in a pipeline.
```

File: lib/elixir/lib/kernel.ex (hunk 3 above)
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:1141]:
```suggestion
      iex> tap(1, fn x -> x + 1 end)
      1

  Most commonly, this is used in pipelines. For example,
  let's suppose you want to inspect part of a data structure.
  You could write:

      %{a: 1}
      |> Map.update!(:a, & &1 + 2)
      |> tap(&IO.inspect(&1.a))
      |> Map.update!(:a, & &1 * 2)

```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**